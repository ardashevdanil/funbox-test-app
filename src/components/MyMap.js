import PropTypes from 'prop-types';
import React from 'react';
import { YMaps, Map, Polyline } from 'react-yandex-maps';

import MyPlacemark from './MyPlacemark';

class MyMap extends React.Component {
  constructor(props) {
    super(props);
    this.mapRef = null;
    this.mapState = { center: [55.76, 37.64], zoom: 10 };
    this.handlePlacemarkDrag = this.handlePlacemarkDrag.bind(this);
  }

  handlePlacemarkDrag(e, id) {
    //converting page pixel coords to global map coords
    const coords = this.mapRef.options.get('projection').fromGlobalPixels(
      this.mapRef.converter.pageToGlobal(e.get('position')),
      this.mapRef.getZoom()
    );
    
    this.props.setPointCoords(coords, id);
  }

  componentDidUpdate(prevProps) {
    //setting the point coords after creating
    if (this.props.points.length > prevProps.points.length) {
      const pointId = this.props.points.length - 1;

      this.props.setPointCoords(this.mapRef.getCenter(), pointId);
    }
  }

  render() {
    const lineCoords = [];
    const points = this.props.points.map( (item, index) => {
      if (item.coords) {
        lineCoords.push(item.coords);

        return (
          <MyPlacemark
            key={item.key}
            onDrag={this.handlePlacemarkDrag}
            coords={item.coords}
            value={item.value}
            index={index}
          />
        );
      }
    });

    return (
      <YMaps>
        <Map
          width='50%'
          height='100%'
          state={this.mapState}
          instanceRef={(elem) => this.mapRef = elem}
        >
          {points}
          <Polyline
            geometry={{
              coordinates: lineCoords
            }}
            options={{
              strokeColor: '#FF00FF',
              strokeWidth: 4
            }}
          />
        </Map>
      </YMaps>
    );
  }
};

MyMap.propTypes = {
  points: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
      ]),
      key: PropTypes.number,
      position: PropTypes.number,
      coords: PropTypes.arrayOf(PropTypes.number),
    })
  ).isRequired,
  setPointCoords: PropTypes.func.isRequired,
}

export default MyMap;