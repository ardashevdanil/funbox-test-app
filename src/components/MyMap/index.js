import PropTypes from 'prop-types';
import React from 'react';
import { Map, Placemark, Polyline, YMaps } from 'react-yandex-maps';
import { compose, withState } from 'recompose';
import { connect } from 'react-redux';

import { setMapCenter } from '../../redux/modules/mapCenter';
import { setPointCoords } from '../../redux/modules/points';

const MyMap = ({
  mapRef,
  mapState,
  points,
  setCoordsAction,
  setMapCenterAction,
  setMapRef,
}) => (
  <YMaps>
    <Map
      instanceRef={(elem) => setMapRef(elem)}
      height='100%'

      // Setting the map center's coords after moving the map
      onBoundsChange={() => setMapCenterAction(mapRef.getCenter())}
      state={mapState}
      width='50%'
    >
      {points.map((point, index) => (
        <Placemark
          geometry={{
            coordinates: point.coords,
            type: 'Point',
          }}
          key={point.key}
          onDrag={e => {
            // Converting page pixel coords to global map coords
            const coords = mapRef.options.get('projection').fromGlobalPixels(
              mapRef.converter.pageToGlobal(e.get('position')),
              mapRef.getZoom(),
            );
            setCoordsAction({ coords, key: point.key });
          }}
          options={{ draggable: true }}
          properties={{ 
            balloonContent: point.value,
            iconContent: index + 1, 
          }}
        />
      ))}
      <Polyline
        geometry={{
          coordinates: points.map(point => point.coords), 
        }}
      />
    </Map>
  </YMaps>
);

const enhance = compose(
  connect(
    state => ({ points: state.points }),
    { setCoordsAction: setPointCoords, setMapCenterAction: setMapCenter }),
  withState('mapRef', 'setMapRef', null),
  withState('mapState', 'setMapState', { center: [55.76, 37.64], zoom: 10 }),
);

export default enhance(MyMap);

MyMap.propTypes = {
  mapState: PropTypes.object.isRequired,
  points: PropTypes.arrayOf(
    PropTypes.shape({
      coords: PropTypes.arrayOf(PropTypes.number).isRequired,
      key: PropTypes.number.isRequired,
      value: PropTypes.string.isRequired,
    })
  ).isRequired,
  setCoordsAction: PropTypes.func.isRequired,
  setMapCenterAction: PropTypes.func.isRequired,
  setMapRef: PropTypes.func.isRequired,
}
