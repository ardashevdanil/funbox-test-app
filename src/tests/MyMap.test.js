import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';

import MyMap from '../components/MyMap';

Enzyme.configure({ adapter: new Adapter() });

describe('test MyMap', () => {
  const setPointCoords = jest.fn();
	const component = Enzyme.shallow(
    <MyMap
      points={[
        {
          coords: [1, 1],
          key: 1,
          value: 'value',
        }
      ]}
      setPointCoords={setPointCoords}
    />
  );

  test('renders correctly', () => {
    expect(component.debug()).toMatchSnapshot();
  });

  test('pass coords to Polyline', () => {
    expect(component.find('Polyline').props().geometry.coordinates)
      .toEqual([[1, 1]]);
  });

  test('renders placemarks correctly', () => {
    expect(component.find('MyPlacemark').length).toBe(1);
  });
});