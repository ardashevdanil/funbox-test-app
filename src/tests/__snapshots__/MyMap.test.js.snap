// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MyMap renders correctly 1`] = `
"<YMaps enterprise={false} onApiAvaliable={[Function]} version=\\"2.1\\">
  <Map width=\\"50%\\" height=\\"100%\\" state={{...}} instanceRef={[Function: instanceRef]}>
    <MyPlacemark onDrag={[Function: bound handlePlacemarkDrag]} coords={{...}} value=\\"value\\" index={0} />
    <Polyline geometry={{...}} options={{...}} />
  </Map>
</YMaps>"
`;
